charmcraft/commands/build.py

```diff
@@ -147,25 +147,35 @@ def handle_dispatcher(self, linked_entrypoint):
                 fileno = fh.fileno()
                 os.fchmod(fileno, os.fstat(fileno).st_mode | stat.S_IXUSR)
 
-        # bunch of symlinks, to support old juju; whatever is in the charm's hooks directory
-        # is respected, but also the mandatory ones are created if missing
+        # bunch of symlinks, to support old juju: whatever is in the charm's hooks directory
+        # is respected (unless links to the entrypoint), but also the mandatory ones are
+        # created if missing
         current_hookpath = self.charmdir / 'hooks'
         dest_hookpath = self.buildpath / 'hooks'
         dest_hookpath.mkdir()
+
+        # get current hooks, separating those to be respected verbatim, and those that we need
+        # to replace (because they are pointing to the entrypoint and we need to fix the
+        # environment in the middle)
+        current_hooks_ok = []
+        current_hooks_to_replace = []
         if current_hookpath.exists():
-            current_hooks = list(current_hookpath.iterdir())
-        else:
-            current_hooks = []
+            for node in current_hookpath.iterdir():
+                if node.resolve() == self.entrypoint:
+                    current_hooks_to_replace.append(node)
```

Comment:
further to @jameinel's comment about logging, I think something like
```suggestion
                    current_hooks_to_replace.append(node)
                    logger.debug("Ignoring existing hook %r as it's a symlink to the entrypoint", node)
```
would do the trick